ext/drreg/drreg.c

```diff
@@ -173,15 +184,58 @@ get_where_app_pc(instr_t *where)
  * SPILLING AND RESTORING
  */
 
+static ptr_uint_t drreg_note_base = 0;
+
+/* Get note value for a drreg instr note. */
+static ptr_uint_t
+get_drreg_note_val(uint val)
+{
+    ASSERT(drreg_note_base > 0, "drreg_note_base not initialized");
+    return (ptr_uint_t)(drreg_note_base + val);
+}
+
+/* Returns whether the given ilist has an existing usage of the given slot
+ * on or after `where`. Such a usage, if any, must have been added by a
+ * previous instrumentation pass, and tells us that it is not safe to reuse
+ * the slot in the current pass.
+ */
+static bool
+has_pending_slot_usage_by_prior_pass(per_thread_t *pt, instrlist_t *ilist, instr_t *where,
+                                     uint slot)
+{
+    if (!TEST(DRREG_HANDLE_MULTI_PHASE_SLOT_RESERVATIONS, pt->bb_props))
+        return false;
+    for (instr_t *in = where; in != NULL /*sanity*/ && in != instrlist_last(ilist);
+         in = instr_get_next(in)) {
+        /* We store data about spill slot usage in the data area of a label instr
+         * immediately preceding the usage.
+         */
+        if (instr_is_label(in) &&
+            instr_get_note(in) == (void *)get_drreg_note_val(DRREG_NOTE_SPILL_SLOT_ID)) {
+            dr_instr_label_data_t *data = instr_get_label_data_area(in);
+            ASSERT(data != NULL, "failed to find label's data area");
+            /* Try to continue in release build. */
+            if (data != NULL &&
+                data->data[SPILL_SLOT_ID_LABEL_INSTR_DATA_AREA_OFFSET] ==
```

Comment:
I found a bit late that the existing `drreg_is_instr_spill_or_restore`[1] may be useful for doing this check. Though I don't think using `drreg_is_instr_spill_or_restore` is ideal -- it documents that it "may not be able to distinguish a drreg spill or restore from some other spill or restore", and also there may be more overhead in looking at each instr using `drreg_is_instr_spill_or_restore` vs just looking at label instrs with more readily available spill slot usage data.

[1]: https://github.com/DynamoRIO/dynamorio/blob/dcbaca81b9816a40fddc2b0211475a7c5a9220a3/ext/drreg/drreg.h#L610

Reply:
This is what I was referring to with the earlier comment and the reference to #3801 where we want to replace the from-cache decode with the recreate IR metadata, for cases of xl8-by-recreate.

Reply:
I would suggest putting a comment about using this metadata with #3801 in the drreg restore state code.

Reply:
Done.